samples/nrf_rpc/protocols_serialization/client/src/bt_test_shell.c

```diff
@@ -0,0 +1,281 @@
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(bt_test, CONFIG_NRF_PS_CLIENT_LOG_LEVEL);
+
+/* UUID: 12345678-1234-5678-1234-56789abcdef0 */
+#define BT_UUID_CUSTOM_SERVICE_VAL                                                                 \
+	BT_UUID_128_ENCODE(0x12345678, 0x1234, 0x5678, 0x1234, 0x56789abcdef0)
+#define BT_UUID_CUSTOM_CHAR_VAL                                                                    \
+	BT_UUID_128_ENCODE(0xabcdef01, 0x2345, 0x6789, 0xabcd, 0xef0123456789)
+
+#ifdef CONFIG_NRF_PS_CLIENT_BT_TEST_DATA_SIZE
+#define TEST_DATA_SIZE (CONFIG_NRF_PS_CLIENT_BT_TEST_DATA_SIZE)
+#else
+#define TEST_DATA_SIZE (10)
+#endif
+
+static void connected(struct bt_conn *conn, uint8_t hci_err);
+static void disconnected(struct bt_conn *conn, uint8_t reason);
+static void update_subscribed(void);
+static void ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value);
+static ssize_t read_custom_char(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+				uint16_t len, uint16_t offset);
+static ssize_t write_custom_char(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				 const void *buf, uint16_t len, uint16_t offset, uint8_t flags);
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR),
+	BT_DATA_BYTES(BT_DATA_UUID128_ALL, BT_UUID_CUSTOM_SERVICE_VAL),
+};
+
+static const struct bt_data sd[] = {
+	BT_DATA(BT_DATA_NAME_COMPLETE, CONFIG_BT_DEVICE_NAME, sizeof(CONFIG_BT_DEVICE_NAME) - 1),
+};
+
+static struct bt_uuid_128 custom_service_uuid = BT_UUID_INIT_128(BT_UUID_CUSTOM_SERVICE_VAL);
+static struct bt_uuid_128 custom_char_uuid = BT_UUID_INIT_128(BT_UUID_CUSTOM_CHAR_VAL);
+
+static struct bt_gatt_indicate_params ind_params;
+
+static struct bt_conn_cb conn_cb = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
+
+typedef struct {
+	uint8_t data[TEST_DATA_SIZE];
+	bool notify_enabled;
+	bool indicate_enabled;
+	struct bt_conn *conn;
+} test_svc_ctx_t;
+
+static test_svc_ctx_t test_ctx = {
+	.data = {},
+	.notify_enabled = false,
+	.indicate_enabled = false,
+	.conn = NULL,
+};
+
+BT_GATT_SERVICE_DEFINE(custom_svc, BT_GATT_PRIMARY_SERVICE(&custom_service_uuid),
+		       BT_GATT_CHARACTERISTIC(&custom_char_uuid.uuid,
+					      BT_GATT_CHRC_READ | BT_GATT_CHRC_WRITE |
+						      BT_GATT_CHRC_NOTIFY | BT_GATT_CHRC_INDICATE,
+					      BT_GATT_PERM_READ | BT_GATT_PERM_WRITE,
+					      read_custom_char, write_custom_char, &test_ctx),
+		       BT_GATT_CCC(ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE));
+
+static ssize_t read_custom_char(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+				uint16_t len, uint16_t offset)
+{
+	LOG_INF("BT Test Read. off %d len %d", offset, len);
+
+	if (offset >= TEST_DATA_SIZE) {
+		LOG_ERR("BT Test Invalid offset");
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	test_svc_ctx_t *ctx = attr->user_data;
+	uint16_t actual_len = MIN(len, TEST_DATA_SIZE - offset);
```

Comment:
Please declare the var at the beginning. I think clang warns about it.

Reply:
Done